File: benchmark/nested_loop.py
#!/usr/bin/env python3
import os
import time

N = int(os.environ.get("BENCH_NESTED_N", "520"))

def nested_loop() -> int:
    acc = 0
    for i in range(N):
        for j in range(N):
            for k in range(N):
                acc += (i ^ j ^ k) & 0xFF
    return acc

def main() -> None:
    t0 = time.perf_counter()
    result = nested_loop()
    ms = int((time.perf_counter() - t0) * 1000)
    print(f"{'Python':<8} {'nested_loop':<14} {ms:8d} ms  result={result}")

if __name__ == "__main__":
    main()

File: benchmark/nested_loop.c
#include "bench_util.h"

int main(void) {
  const int64_t n = bench_env_i64("BENCH_NESTED_N", 520);
  int64_t acc = 0;
  int64_t t0 = bench_now_ms();
  for (int64_t i = 0; i < n; ++i) {
    for (int64_t j = 0; j < n; ++j) {
      for (int64_t k = 0; k < n; ++k) {
        acc += (i ^ j ^ k) & 0xff;
      }
    }
  }
  int64_t ms = bench_now_ms() - t0;
  bench_print_line("C", "nested_loop", ms, (uint64_t)acc);
  return 0;
}
